modules/server/src/main/scala/scaladex/server/service/UserSessionSynchronizer.scala

```diff
@@ -0,0 +1,65 @@
+package scaladex.server.service
+
+import java.util.UUID
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model.StatusCodes
+import cats.implicits.toTraverseOps
+import scaladex.core.model.GithubResponse
+import scaladex.core.model.UserInfo
+import scaladex.core.model.UserState
+import scaladex.core.service.GithubService
+import scaladex.core.service.SchedulerDatabase
+import scaladex.infra.GithubClient
+
+class UserSessionSynchronizer(database: SchedulerDatabase)(implicit ec: ExecutionContext, ac: ActorSystem)
+    extends Scheduler("user-session-synchronizer", 1.hour) {
+
+  type Session = (UUID, UserState)
+
+  override def run(): Future[Unit] =
+    for {
+      sessions <- database.getAllSessions()
+      updatedEitherUserSessions <- sessions.traverse { case (session, client) => updateUserSession(session, client) }
+      expiredSessionIds = updatedEitherUserSessions.collect { case Left(expiredSessionId) => expiredSessionId }
+      updatedUserSessions = updatedEitherUserSessions.collect { case Right(sessions) => sessions }
+      _ <- updatedUserSessions.traverse { case (userId, userState) => database.insertSession(userId, userState) }
+      _ <- expiredSessionIds.traverse(database.deleteSession)
+    } yield ()
+
+  private def updateUserSession(session: Session): Future[Either[UUID, Session]] =
+    session match {
+      case (userId, staleUserState) =>
+        val service = new GithubClient(staleUserState.info.token)
+        getUserInfo(userId, staleUserState.info, service)
+          .map(_.map { updatedUserInfo =>
+            if (updatedUserInfo != staleUserState.info)
+              (userId, staleUserState.copy(info = updatedUserInfo))
+            else session
+          })
+    }
+
+  private def getUserInfo(
+      userId: UUID,
+      staleUserInfo: UserInfo,
+      service: GithubService
+  ): Future[Either[UUID, UserInfo]] =
+    service.getUserInfo().flatMap {
```

Comment:
Sorry I should have catched this earlier: we also need to update the `repos` and `organizations` in `UserState`.

So maybe you should add a `getUserState` method in `GithubService` and call it instead of `getUserInfo`. There is one already in `GithubAuthImpl` that you can move to `GithubClient`.

Reply:
That sounds good. Here's what I have so far

```scala
  override def getUserState(token: Secret): Future[GithubResponse[UserState]] = {
    val permissions = Seq("WRITE", "MAINTAIN", "ADMIN")
    (for {
      response <- getUserInfo()
      userInfo <- response match {
        case GithubResponse.Ok(userInfo)               => Future.successful(userInfo)
        case GithubResponse.MovedPermanently(userInfo) => Future.successful(userInfo)
        case GithubResponse.Failed(code, errorMessage) =>
          Future.failed(new Exception(s"Request to GithubClient#getUserInfo failed with code: $code, reason: $errorMessage"))
      }
      userOrgs <- getUserOrganizations(userInfo.login).recover(_ => Seq.empty)
      userOrgRepos <- userOrgs
        .traverse(getOrganizationRepositories(userInfo.login, _, permissions))
        .recover(_ => Seq.empty)
        .map(_.flatten)
      userRepos <- getUserRepositories(userInfo.login, permissions).recover(_ => Seq.empty)
    } yield GithubResponse.Ok(UserState(userOrgRepos.toSet ++ userRepos, userOrgs.toSet, userInfo)))
      .recoverWith {
        // TODO: Figure out how to assign the correct code.
        case t: Throwable => GithubResponse.Failed()
      }
  }
```

As you can see, I'm having some issues mapping the error in the for-comprehension into the `.recoverWith { ... }` at the end. Maybe I'm overthinking this and this isn't that hard... but I think it's hard to get information about what specific type of error that was thrown.

I could imagine making the result of a the for-comp an `Either[A, UserState]`, where `A` contains specific information about the type of failure, and then transforming the `Either[A, UserState]` into a `GithubResponse[UserState]` at the end, but this seems like overkill... any thoughts?

Reply:
I got this working


```scala
  override def getUserState(token: Secret): Future[GithubResponse[UserState]] = {
    val permissions = Seq("WRITE", "MAINTAIN", "ADMIN")
    for {
      response <- getUserInfo()
      eitherInfo = response match {
        case GithubResponse.Ok(userInfo)               => userInfo.asRight
        case GithubResponse.MovedPermanently(userInfo) => userInfo.asRight
        case failed: GithubResponse.Failed             => failed.asLeft
      }
      eitherUserOrgs <- eitherInfo.traverse(info => getUserOrganizations(info.login).recover(_ => Seq.empty))
      eitherUserOrgRepos <- (eitherUserOrgs, eitherInfo).traverseN { (userOrgs, info) =>
        userOrgs
          .traverse(getOrganizationRepositories(info.login, _, permissions))
          .map(_.flatten)
      }
      eitherUserRepos <- eitherInfo.traverse(info =>
        getUserRepositories(info.login, permissions).recover(_ => Seq.empty)
      )
    } yield (eitherInfo, eitherUserOrgs, eitherUserOrgRepos, eitherUserRepos)
      .mapN { (info, userOrgs, userOrgRepos, userRepos) =>
        GithubResponse.Ok(UserState(repos = userOrgRepos.toSet ++ userRepos, orgs = userOrgs.toSet, info = info))
      }
      .fold(identity, identity)
  }
```

But maybe it's not the most readable? with `.traverse`, `.traverseN`, and `.mapN`?